### PR TITLE
[API][Order] Orders browsing, cancelling and cancelling with promotion

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | api
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        );
+    }
+
+    /**
+     * @Then I should see a single order from customer :customer
+     */
+    public function iShouldSeeASingleOrderFromCustomer(CustomerInterface $customer): void
+    {
+        Assert::true($this->responseChecker->hasItemWithValue(
+            $this->client->getLastResponse(),
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
            $this->client->index(),
```

/cc @GSadee WDYT?

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+    /**
+     * @Then this order should have state :state
+     * @Then its state should be :state
+     */
+    public function itsStateShouldBe(string $state): void
+    {
+        Assert::true(
+            $this->responseChecker->hasValue(
+                $this->client->show($this->sharedStorage->get('order')->getNumber()),
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: I'm not sure, if we shouldn't use iri converter and a whole object here

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+            sprintf('payment for this order is not %s', $state)
+        );
+    }
+
+    /**
+     * @Then /^there should be(?:| only) (\d+) payments?$/
+     */
+    public function theOrderShouldHaveNumberOfPayments(int $number): void
+    {
+        Assert::eq(count($this->responseChecker->getValue($this->client->show($this->sharedStorage->get('order')->getNumber()), 'payments')), $number);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
        Assert::count($this->responseChecker->getValue($this->client->show($this->sharedStorage->get('order')->getNumber()), 'payments'), $number);
```

What is more, this line is too long. Can we extract `$this->sharedStorage->get('order')->getNumber()` to separate variable.

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php (hunk 1 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: I makes even more sense, when you check lines like https://github.com/Sylius/Sylius/pull/11239/files#diff-b3bba723ddead4c0bc89c6fe061844feR150 where we are not using last response but requesting order state from the scratch.

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        $this->client->show($order->getNumber());
+    }
+
+    /**
+     * @When I cancel this order
+     */
+    public function iCancelThisOrder(): void
+    {
+        $this->client->applyTransition(
+            $this->responseChecker->getValue($this->client->getLastResponse(), 'number'),
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: What will happen, if the previous message didn't contain order? I would recommend using `this order` to create a new uri.

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    {
+        Assert::same(
+            $this->responseChecker->getValue($this->client->show($this->sharedStorage->get('order')->getNumber()), 'orderPromotionTotal'),
+            $this->getPriceFromString($promotionTotal)
+        );
+    }
+
+    private function getPriceFromString(string $price): int
+    {
+        return (int) round((float) str_replace(['€', '£', '$'], '', $price) * 100, 2);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: Should be handled by lexical context

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php (hunk 1 above)
Comment (AdamKasp) [reply src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]: yes, but in this way, step `When I browse orders` is redundant in api cases, and step `Then I should see a single order from customer "john.doe@gmail.com"` will work only if we have simple browsing. What about when we check filtering response? in one step we filtering and in next step we checking filtered data. Am I right?

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    {
+        $this->iCancelThisOrder($order);
+        Assert::contains(
+            $this->responseChecker->getError($this->client->getLastResponse()),
+            'Transition "cancel" cannot be applied on state "cancelled"'
+        );
+    }
+
+    /**
+     * @Then /^the order's total should(?:| still) be "(\-)?(?:€|£|￥|\$)((?:\d+\.)?\d+)"$/
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
     * @Then /^the order's total should(?:| still) be "([^"]+)"$/
```
As in src/Sylius/Behat/Context/Ui/Admin/ManagingOrdersContext.php:317

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+    /**
+     * @Then this order should have state :state
+     * @Then its state should be :state
+     */
+    public function itsStateShouldBe(string $state): void
+    {
+        Assert::true(
+            $this->responseChecker->hasValue(
+                $this->client->show($this->sharedStorage->get('order')->getNumber()),
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
                $this->client->getLastResponse(),
```
And same usage for everything that refers to the current context?

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        );
+    }
+
+    /**
+     * @Then /^the order's total should(?:| still) be ("[^"]+")$/
+     */
+    public function theOrdersTotalShouldBe(int $total): void
+    {
+        Assert::same(
+            $this->responseChecker->getValue($this->client->show($this->sharedStorage->get('order')->getNumber()), 'total'),
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
            $this->responseChecker->getValue($this->client->getLastResponse(), 'total'),
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        );
+    }
+
+    /**
+     * @Then /^the order's promotion total should(?:| still) be ("[^"]+")$/
+     */
+    public function theOrdersPromotionTotalShouldBe(int $promotionTotal): void
+    {
+        Assert::same(
+            $this->responseChecker->getValue($this->client->show($this->sharedStorage->get('order')->getNumber()), 'orderPromotionTotal'),
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingOrdersContext.php]:
```suggestion
            $this->responseChecker->getValue($this->client->getLastResponse(), 'orderPromotionTotal'),
```

**Comment by lchrusciel** [comment]
Thank you, Adam! :1st_place_medal:

**Comment by lchrusciel** [comment]
Part of #11250

**Pull request merged**